batch/batch/driver/canceller.py

```diff
@@ -94,39 +94,44 @@ async def cancel_cancelled_ready_jobs_loop_body(self):
         }
 
         async def user_cancelled_ready_jobs(user, remaining) -> AsyncIterator[Dict[str, Any]]:
-            async for batch in self.db.select_and_fetchall(
+            async for job_group in self.db.select_and_fetchall(
                 """
-SELECT batches.id, job_groups_cancelled.id IS NOT NULL AS cancelled
-FROM batches
-LEFT JOIN job_groups_cancelled
-       ON batches.id = job_groups_cancelled.id
+SELECT job_groups.batch_id, job_groups.job_group_id, t.cancelled IS NOT NULL AS cancelled
+FROM job_groups
+LEFT JOIN LATERAL (
+  SELECT 1 AS cancelled
+  FROM job_group_self_and_ancestors
+  INNER JOIN job_groups_cancelled
+    ON job_group_self_and_ancestors.batch_id = job_groups_cancelled.id AND
+      job_group_self_and_ancestors.ancestor_id = job_groups_cancelled.job_group_id
+  WHERE job_groups.batch_id = job_group_self_and_ancestors.batch_id AND
+    job_groups.job_group_id = job_group_self_and_ancestors.job_group_id
+) AS t ON TRUE
```

Comment:
You've got it backwards. Traversing the tree upwards is cheap. If we make the max depth 5, then that's 5 records. However, traversing down the tree is slow. Let's say we're trying to cancel the root job group and there's a massive tree with 100K direct children job groups and each of those has additional children. Then this INSERT statement to insert all of the records at once is going to be extremely slow. It's the same reason why we don't just update all of the jobs records at once when we cancel and have these checks on whether the job group has been cancelled.

With regards to all of the checks about ancestors being cancelled, yes, I tried to put them in all the correct places.

Please let me know if you disagree with this reasoning as the implication of it changes a lot of your other comments and code in other places.

Reply:
We might be able to do this with a SQL function. Then it would just be `is_cancelled(job_groups.batch_id, job_groups.job_group_id) AS cancelled`.

```
SELECT batches.*,
  is_cancelled(job_groups.batch_id, job_groups.job_group_id) AS cancelled,
  job_groups_n_jobs_in_complete_states.n_completed,
  job_groups_n_jobs_in_complete_states.n_succeeded,
  job_groups_n_jobs_in_complete_states.n_failed,
  job_groups_n_jobs_in_complete_states.n_cancelled,
  cost_t.*
FROM job_groups
LEFT JOIN batches ON batches.id = job_groups.batch_id
LEFT JOIN job_groups_n_jobs_in_complete_states
       ON job_groups.batch_id = job_groups_n_jobs_in_complete_states.id AND job_groups.job_group_id = job_groups_n_jobs_in_complete_states.job_group_id
LEFT JOIN LATERAL (
  SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
  FROM (
    SELECT resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
    FROM aggregated_job_group_resources_v3
    WHERE job_groups.batch_id = aggregated_job_group_resources_v3.batch_id AND job_groups.job_group_id = aggregated_job_group_resources_v3.job_group_id
    GROUP BY resource_id
  ) AS usage_t
  LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
) AS cost_t ON TRUE
WHERE job_groups.batch_id = %s AND job_groups.job_group_id = %s AND NOT deleted;
```

Reply:
I tried replacing all of these lateral joins with a stored function and it mostly worked except there was some sort of deadlock in the `cancel_job_group` stored procedure. I think we should make an issue to fix this later on and replace the use of the lateral joins with an `is_job_group_cancelled` function as a separate project. At least we know the lateral joins should perform well and I wasn't 100% convinced the functions had the same performance.